File: web/services/feed_catalog.py
"""Feed catalog service — loads and queries the curated RSS feed list."""

import os
from functools import lru_cache

import yaml


CATALOG_PATH = os.path.join(os.path.dirname(__file__), "..", "data", "feed_catalog.yaml")


@lru_cache(maxsize=1)
def _load_catalog() -> dict:
    """Load the feed catalog from YAML."""
    with open(CATALOG_PATH) as f:
        return yaml.safe_load(f)


def get_bundles() -> list[dict]:
    """Get all starter bundles.

    Returns a list of dicts with keys: name, description, feeds (list of feed IDs).
    """
    return _load_catalog()["bundles"]


def get_categories() -> list[dict]:
    """Get all feed categories with their feeds.

    Returns a list of dicts with keys: name, feeds (list of feed dicts).
    Each feed dict has: id, name, url, description.
    """
    return _load_catalog()["categories"]


def get_all_feeds() -> dict[str, dict]:
    """Get a flat dict of all feeds, keyed by feed ID.

    Returns: {"guardian-world": {"id": "guardian-world", "name": "The Guardian", ...}, ...}
    """
    feeds = {}
    for category in get_categories():
        for feed in category["feeds"]:
            feeds[feed["id"]] = feed
    return feeds


def get_feeds_for_bundle(bundle_name: str) -> list[dict]:
    """Get the full feed details for a given bundle name.

    Returns a list of feed dicts (id, name, url, description).
    """
    all_feeds = get_all_feeds()
    for bundle in get_bundles():
        if bundle["name"] == bundle_name:
            return [all_feeds[fid] for fid in bundle["feeds"] if fid in all_feeds]
    return []


def describe_feed_selection(feed_urls: set[str]) -> str:
    """Describe a feed selection as bundles + extras.

    Examples:
        "Morning Briefing"
        "Morning Briefing + 3 extra sources"
        "Morning Briefing and Tech & Science"
        "Morning Briefing, Tech & Science, and Business & Finance + 1 extra source"
        "15 sources"  (no complete bundles matched)
    """
    matched_bundles = []
    covered_urls: set[str] = set()
    for bundle in get_bundles():
        bundle_feeds = get_feeds_for_bundle(bundle["name"])
        if bundle_feeds and all(f["url"] in feed_urls for f in bundle_feeds):
            matched_bundles.append(bundle["name"])
            for f in bundle_feeds:
                covered_urls.add(f["url"])

    extra_count = len(feed_urls - covered_urls)

    if not matched_bundles:
        count = len(feed_urls)
        return f"{count} source{'s' if count != 1 else ''}"

    if len(matched_bundles) == 1:
        bundle_text = matched_bundles[0]
    elif len(matched_bundles) == 2:
        bundle_text = f"{matched_bundles[0]} and {matched_bundles[1]}"
    else:
        bundle_text = ", ".join(matched_bundles[:-1]) + f", and {matched_bundles[-1]}"

    if extra_count > 0:
        return f"{bundle_text} + {extra_count} extra source{'s' if extra_count != 1 else ''}"
    return bundle_text


def validate_rss_url(url: str) -> bool:
    """Basic validation of an RSS feed URL."""
    if not url:
        return False
    url = url.strip()
    if not url.startswith(("http://", "https://")):
        return False
    if "." not in url:
        return False
    return True

File: web/src/lib/feed-catalog.ts
import yaml from "js-yaml";
import fs from "fs";
import path from "path";
import type { CatalogBundle, CatalogCategory, CatalogFeed } from "@/types";

interface CatalogData {
  bundles: {
    name: string;
    description: string;
    feeds: string[];
  }[];
  categories: {
    name: string;
    feeds: {
      id: string;
      name: string;
      url: string;
      description: string;
    }[];
  }[];
}

let catalog: CatalogData | null = null;

function loadCatalog(): CatalogData {
  if (catalog) return catalog;
  const filePath = path.join(process.cwd(), "src/data/feed-catalog.yaml");
  const raw = fs.readFileSync(filePath, "utf-8");
  catalog = yaml.load(raw) as CatalogData;
  return catalog;
}

export function getBundles(): CatalogBundle[] {
  const data = loadCatalog();
  return data.bundles.map((b) => ({
    name: b.name,
    description: b.description,
    feeds: b.feeds,
  }));
}

export function getCategories(): CatalogCategory[] {
  const data = loadCatalog();
  return data.categories.map((c) => ({
    name: c.name,
    feeds: c.feeds.map((f) => ({
      id: f.id,
      name: f.name,
      url: f.url,
      description: f.description,
    })),
  }));
}

export function getAllFeeds(): Map<string, CatalogFeed> {
  const data = loadCatalog();
  const map = new Map<string, CatalogFeed>();
  for (const category of data.categories) {
    for (const feed of category.feeds) {
      map.set(feed.id, {
        id: feed.id,
        name: feed.name,
        url: feed.url,
        description: feed.description,
      });
    }
  }
  return map;
}

export function getFeedsForBundle(bundleName: string): CatalogFeed[] {
  const bundles = getBundles();
  const bundle = bundles.find((b) => b.name === bundleName);
  if (!bundle) return [];

  const allFeeds = getAllFeeds();
  return bundle.feeds
    .map((id) => allFeeds.get(id))
    .filter((f): f is CatalogFeed => f !== undefined);
}

/**
 * Human-readable description of the user's feed selection.
 * Mirrors web/services/feed_catalog.py::describe_feed_selection()
 */
export function describeFeedSelection(feedUrls: Set<string>): string {
  if (feedUrls.size === 0) return "No sources selected";

  const bundles = getBundles();
  const allFeeds = getAllFeeds();

  // Build a set of URLs per bundle
  const bundleUrlSets = new Map<string, Set<string>>();
  for (const bundle of bundles) {
    const urls = new Set<string>();
    for (const feedId of bundle.feeds) {
      const feed = allFeeds.get(feedId);
      if (feed) urls.add(feed.url);
    }
    bundleUrlSets.set(bundle.name, urls);
  }

  // Find which bundles are fully contained in the selection
  const matchedBundles: string[] = [];
  const coveredUrls = new Set<string>();
  for (const [name, urls] of bundleUrlSets) {
    const allIncluded = [...urls].every((url) => feedUrls.has(url));
    if (allIncluded) {
      matchedBundles.push(name);
      for (const url of urls) coveredUrls.add(url);
    }
  }

  const extraCount = feedUrls.size - coveredUrls.size;

  if (matchedBundles.length === 0) {
    return `${feedUrls.size} source${feedUrls.size === 1 ? "" : "s"}`;
  }

  const bundleStr = matchedBundles.join(" and ");
  if (extraCount > 0) {
    return `${bundleStr} + ${extraCount} extra source${extraCount === 1 ? "" : "s"}`;
  }
  return bundleStr;
}

/**
 * Basic RSS URL validation.
 * Mirrors web/services/feed_catalog.py::validate_rss_url()
 */
export function validateRssUrl(url: string): boolean {
  if (!url) return false;
  const trimmed = url.trim().toLowerCase();
  return (
    (trimmed.startsWith("http://") || trimmed.startsWith("https://")) &&
    trimmed.includes(".")
  );
}
